### PR TITLE
Part 1 of Aligning Core Blockchain Package with Latest 2.1 Spec

### PR DESCRIPTION
**Pull request by terenc3t**
First part of aligning beacon chain node code base with latest 2.1 spec. Following items got changed in this PR with respect to 2.1:

- Updates to state and crystallized state fields
- Minor tweaks on shuffling and cutoff algorithms 
- How we apply casper FFG rewards and penalties 

I don't want to put all the 2.1 changes in a single giant PR. Will be opening multiple small in regards to more 2.1 updates. Per blocking processing, cycle length recalculating, test coverage improvements will be coming next.

https://notes.ethereum.org/SCIg8AH5SA-O4C1G1LYZHQ?view

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -354,33 +363,88 @@ func (b *BeaconChain) calculateRewardsFFG() error {
...
+		slotsPerCommittee = 1
+	} else {
+		committeesPerSlot = 1
+		slotsPerCommittee = 1
+		for len(indices)*slotsPerCommittee < params.MinCommiteeSize && slotsPerCommittee < params.CycleLength {
+			slotsPerCommittee *= 2
+		}
+	}
+
+	// split the shuffled list for heights
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: punctuation

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -211,11 +211,17 @@ func (c *ChainService) run(done <-chan struct{}) {
 				log.Errorf("Write active state to disk failed: %v", err)
 			}
 
-			currentSlot := block.SlotNumber()
+			// TODO: Apply 2.1 fork choice logic using the following
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: .

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -36,6 +36,11 @@ type beaconState struct {
 	CrystallizedState *types.CrystallizedState
 }
```
[the remark this answers is not in the document]
Comment (terenc3t) [reply beacon-chain/blockchain/core.go]: @rauljordan thanks for the review! it's ready again!

**Comment by rauljordan** [comment]
I approve but it looks like we'll have to get coverage up by a lot in the blockchain package as it seems it has a lot of stuff untested. Maybe we can revisit that after the spec is a little bit more solid.

**Comment by terenc3t** [comment]
I'll open a PR to improve the coverage tomorrow

**Pull request merged**